libamqpprox/amqpprox_dataratelimitmanager.cpp

```diff
@@ -0,0 +1,83 @@
+#include <amqpprox_dataratelimitmanager.h>
+
+#include <limits>
+#include <string>
+
+namespace Bloomberg {
+namespace amqpprox {
+
+DataRateLimitManager::DataRateLimitManager()
+: d_vhostDataRateQuota()
+, d_vhostDataRateAlarmQuota()
+, d_defaultDataRateQuota(std::numeric_limits<std::size_t>::max())
+, d_defaultDataRateAlarmQuota(std::numeric_limits<std::size_t>::max())
+{
+}
+
+std::size_t
+DataRateLimitManager::getDataRateLimit(const std::string &vhostName)
+{
+    auto vhostRate = d_vhostDataRateQuota.find(vhostName);
+    if (vhostRate == d_vhostDataRateQuota.end()) {
+        return d_defaultDataRateQuota;
+    }
+
+    return vhostRate->second;
+}
+
+std::size_t
+DataRateLimitManager::getDataRateAlarm(const std::string &vhostName)
+{
+    auto vhostRate = d_vhostDataRateAlarmQuota.find(vhostName);
+    if (vhostRate == d_vhostDataRateAlarmQuota.end()) {
+        return d_defaultDataRateAlarmQuota;
+    }
+
+    return vhostRate->second;
+}
+
+std::size_t DataRateLimitManager::getDefaultDataRateLimit()
+{
+    return d_defaultDataRateQuota;
+}
+
+std::size_t DataRateLimitManager::getDefaultDataRateAlarm()
+{
+    return d_defaultDataRateAlarmQuota;
+}
+
+void DataRateLimitManager::setDefaultDataRateLimit(std::size_t quota)
+{
+    d_defaultDataRateQuota = quota;
+}
+
+void DataRateLimitManager::setDefaultDataRateAlarm(std::size_t quota)
```

Comment:
Since the counters which tick down are stored individually under each Session, and changing the values already propagates down into the session, we only need to store what values they would get assigned when they're created here. For this we can just store the default values, plus any vhost overrides which exist.